### ISSUE TITLE
Feature Request – normalise filenames for shell (linux, et. al.) / avoid malicious names (e.g. code injection)

### ISSUE THREAD

**Issue by DirkSchmitt** [quodlibet/quodlibet#2689]
Description
-----------

In the *file renaming* dialog are nice options like ' ' → '_' or remove non-windows chars.
It would be a nice improvement to have option for cleaning up chars, which are needs quoting on a unix shell like bash.

There are some chars like e.g. '&' which have in *bash* (or other common unix shells) a special meaning.
The use of such filenames on shell level (or in a shell script) requires quoting.

In addition there are some other characters which may have a special meaning in some other context.

Benefit
-------

Think about a filename like `a $(rm --force --recursive ~) shell code injection.mp3`

Proposed transformations
------------------------

### replace ###
- & → +  – fork to background

### replace with ' ' ###
- ( and ) – used for test conditions or command grouping
- [ and ] – used for test conditions
- { and } – command grouping
- ;       – close command sequence
- \       – escape char
- $       – shell variables
- ~       – home directory reference

### additional transformations to ' ' ###

In addition it is a good idea to drop some additional chars whith special meaning.

- % – used for urls to encode special charactes (e.g. a ' ')
- : – may be used to specify a „protocol“
- / – directory separator

Post scriptum
-------------
If I missed some chars – so please add them to the *feature request*.

**Comment by lazka** [comment]
Hm, in which situation would you eval a filename in bash and why?

**Comment by DirkSchmitt** [comment]
The point is not that I call `eval` on a filename.
Hopefully my scripts to maintain my personal music repository are save against this misuse.
Around this „missing feature“ I have worked around in my personal „toolchain“.

But it is not unlikley that an arbitrary user may invoke commands like:

- `find mp3Dir -name "*mp3" -exec …`
- `for f in $(find …) ; do … ; done`

I marked this issue in the title as *„feature request“* to normalise file names in a unixoid environment.
The protection against malicious filenames is a addon benifit.

So please don't focus on this aspect.

In short: The feature request is a „normalisation“ of filename from the viewpoint „unix shell like bash“.

**Comment by lazka** [comment]
ok, if this gets made a plugin, I guess why not.

But assuming "it is not unlikley that an arbitrary user may invoke commands like":

What makes you think that a user that doesn't know how to quote shell arguments will then figure out to sanitize his/her music library, find this option, and then recognize that it is the right solution? That seems unlikely to me.

**Comment by DirkSchmitt** [comment]
I assume that I'm not the only one using exfalso as the tool for sort and put away the „incoming“ music files.
So It would be useful to have such a „switch“ in addition to  the others already available in the „rename files“ tab:
- „Replace spaces with underscores“
- „Strip WIndows-incomapatible characters"
- „Strip diacritical marks“
- „Strip non-ASCII characters“
- „Use only lowercase characters“

**Comment by sauyon** [comment]
I would just like to note that `find`, `xargs`, and `parallel` all deal with these kinds of filenames perfectly reasonably and so the only real issue *is* if somebody tries to `eval` the filenames, as far as I can tell.
